src/services/kratos/auth-email-no-password.ts

```diff
@@ -24,50 +73,133 @@ export const AuthWithEmailPasswordlessService = (): IAuthWithEmailPasswordlessSe
         },
       })
 
-      return data.id
+      return data.id as FlowId
     } catch (err) {
       return new UnknownKratosError(err)
     }
   }
 
-  const validateEmailVerification = async ({
-    code,
-    flow,
-  }: {
-    code: string
-    flow: string
-  }) => {
+  const validateCode = async ({ code, flowId }: { code: EmailCode; flowId: FlowId }) => {
     const method = "code"
 
     try {
       const res = await kratosPublic.updateRecoveryFlow({
-        flow,
+        flow: flowId,
         updateRecoveryFlowBody: {
           method,
           code,
         },
       })
 
-      const wrongCodeMessage =
-        "The recovery code is invalid or has already been used. Please try again."
-      if (!!res.data.ui.messages && res.data.ui.messages[0].text === wrongCodeMessage) {
-        return new PhoneCodeInvalidError()
+      // https://github.com/ory/kratos/blob/b43c50cb8d46638f1b43d4f618dc3631a28cb719/text/id.go#L151
+      const ErrorValidationRecoveryCodeInvalidOrAlreadyUsedId = 4060006
+      if (
+        res.data.ui.messages?.[0].id === ErrorValidationRecoveryCodeInvalidOrAlreadyUsedId
+      ) {
+        return new EmailCodeInvalidError()
       }
 
-      return new UnknownKratosError("happy case should error :/")
+      return new UnknownKratosError("should be a dead branch as 422 error code expected")
     } catch (err) {
-      if (isAxiosError(err) && err.response?.status === 422) {
-        // FIXME bug in kratos? https://github.com/ory/kratos/discussions/2923
-        // console.log("422 response, success?")
-        return true
+      // the recovery flow assume that the user has an additional action
+      // after the code has been verified to reset the password.
+      //
+      // we do not need to do that because we are passwordless
+
+      if (
+        isAxiosError(err) &&
+        err.response?.data?.error?.id === "browser_location_change_required"
+      ) {
+        const cookies: string[] | undefined = err.response?.headers["set-cookie"]
+        if (cookies === undefined) return new UnknownKratosError("cookies undefined")
+
```

Comment:
does this catch path always get called? i set a breakpoint here after running the tests and the `browser_location_change_required` path was always called. are cookies just a requirement for this custom flow?

Reply:
> does this catch path always get called when validating the email code?

yes I think, but only for the happy path. the idea is that in the normal flow you're been askedm at this point, to change your password. but we don't need on our end to complete this flow, as long as we know the user successfully went to this flow, meaning the code was correct, we can authorize the user. 

> are cookies just a requirement for this custom flow?

I think so, until https://github.com/ory/kratos/issues/3163 is been implemented